fabric/src/main/java/me/shedaniel/architectury/mixin/fabric/client/MixinMinecraft.java

```diff
@@ -59,4 +77,78 @@ private void rightClickAir(CallbackInfo ci, InteractionHand var1[], int var2, in
     private void leftClickAir(CallbackInfo ci) {
         InteractionEvent.CLIENT_LEFT_CLICK_AIR.invoker().click(player, InteractionHand.MAIN_HAND);
     }
+    
+    @ModifyVariable(
+            method = "setScreen",
+            at = @At(value = "INVOKE",
+                    target = "Lnet/minecraft/client/player/LocalPlayer;respawn()V",
+                    shift = At.Shift.BY,
+                    by = 2),
+            argsOnly = true
+    )
+    public Screen modifyScreen(Screen screen) {
+        Screen old = screen;
+        InteractionResultHolder<Screen> event = GuiEvent.SET_SCREEN.invoker().modifyScreen(screen);
+        switch (event.getResult()) {
+            case FAIL:
+                setScreenCancelled = true;
+                return old;
+            case SUCCESS:
+                screen = event.getObject();
+                if (old != null && screen != old) {
+                    old.removed();
+                }
+            default:
+                setScreenCancelled = false;
+                return screen;
+        }
+    }
+    
+    @Inject(
+            method = "setScreen",
+            at = @At(value = "INVOKE",
+                    target = "Lnet/minecraft/client/player/LocalPlayer;respawn()V",
+                    shift = At.Shift.BY,
+                    by = 3),
+            cancellable = true
+    )
+    public void cancelSetScreen(@Nullable Screen screen, CallbackInfo ci) {
+        if (setScreenCancelled) {
+            ci.cancel();
+        }
+    }
+    
+    @Redirect(
+            method = "<init>",
+            at = @At(value = "INVOKE", target = "Lnet/minecraft/client/Minecraft;setScreen(Lnet/minecraft/client/gui/screens/Screen;)V"),
+            slice = @Slice(
+                    from = @At(value = "INVOKE", target = "Lnet/minecraft/client/Minecraft;resizeDisplay()V"),
+                    to = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/screens/LoadingOverlay;registerTextures(Lnet/minecraft/client/Minecraft;)V")
+            )
+    )
+    public void minecraftWhy(Minecraft mc, Screen screen) {
```

Comment:
This is the patch I mentioned in my comment that moves setting the initial screen to after resource reloading has occurred.